### PR TITLE
Introduce overrides of protocol members and drop them from witness tables

### PR DESCRIPTION
**Pull request by DougGregor**
When a protocol restates a requirement from its inherited protocol, track that as an override the same way we do with overrides in classes. Such overrides are recorded in the AST but are omitted from the witness table itself. This makes restating protocol requirements in inheriting protocols ABI-neutral, so we can restate protocol requirements to influence associated type inference without baking those decisions into the ABI.

As part of this, introduce tools that allow us to audit all of the restated protocol requirements within the standard library, via a warning new warning flag (`-warn-implicit-overrides`) to warn about such requirements, as well as an appropriate suppression mechanism (`override` to treat it as an override, `@_nonoverride` to give it a new witness table entry). Only those restated requirements that are meant to have a semantic effect, like `BidirectionalCollection.index(_:offsetBy:)`, should use `@_nonoverride`, because it commits us to having that entry in the witness table forever.

Rolling this out in the standard library shrinks the size of the standard library binary by 196k.

Implements rdar://problem/43870489.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
Bah, rebased due to changes on master.

**Comment by DougGregor** [comment]
@airspeedswift  and @moiseev , https://github.com/apple/swift/pull/19034/commits/7fa0b649912b44e8a21c9bcccd876344b7e6d6e6 audits all of the "overriding" protocol requirements to decide which have semantic effect (like `BidirectionalCollection.index(_:offsetBy:)`, marked with `@_nonoverride`) vs. those that are there only to paper over bugs in associated type inference (like `BidirectionalCollection.startIndex`, marked with `override`). I would appreciate a review to verify that I got the set right.

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please benchmark

File: stdlib/public/core/ArrayType.swift [hunk 1]
```diff
@@ -17,13 +17,13 @@ internal protocol ArrayProtocol
 {
   //===--- public interface -----------------------------------------------===//
   /// The number of elements the Array stores.
-  var count: Int { get }
+  override var count: Int { get }
```
Comment (airspeedswift) [inline stdlib/public/core/ArrayType.swift]: Did you add these? Or were they always there, previously unnecessarily?

**Review by airspeedswift — APPROVED** [review]
Your collection ones look right to me. I wonder if there's any theoretical issues like with distance on the integer protocols, but I'm not aware of any.

File: stdlib/public/core/ArrayType.swift (hunk 1 above)
Comment (airspeedswift) [reply stdlib/public/core/ArrayType.swift]: then I suspect these are a weird old anomaly we can remove. Will put up a PR.

Comment (airspeedswift) [reply stdlib/public/core/ArrayType.swift]: (after you land this one to avoid conflicting you)

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: stdlib/public/core/ArrayType.swift (hunk 1 above)
Comment (DougGregor) [reply stdlib/public/core/ArrayType.swift]: Up to you! Once this lands, it's no longer an ABI change to remove it :)

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by jrose-apple** [comment]
Heh heh heh, so this is how `@_nonoverride` gets into the language…

**Comment by DougGregor** [comment]
@jrose-apple yeah, I was thinking that I should be honest about it and pitch the feature

**Comment by DougGregor** [comment]
@airspeedswift the `IterateData` result is interesting. Maybe there's `@_nonoverride` we could add somewhere to improve performance, but I'm not sure if it matters that much.

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by airspeedswift** [comment]
The `IterateDate` benchmark is noisy, it's one of the few that still sometimes crops up on unrelated benchmarks.

**Comment by DougGregor** [comment]
Okay. Assuming I didn't break something w.r.t. the compatibility suite or compile-time performance, I'll go ahead and land this.

**Comment by DougGregor** [comment]
The source-compatibility failure is covered by https://github.com/apple/swift/pull/19146

**Comment by DougGregor** [comment]
From the compiler-performance run:

# Summary for master full

**Unexpected test results, excluded stats for RxSwift, ChattoAdditions, Wordy, ReactiveSwift**

**Regressions found (see below)**

- [Debug-batch](#debug-batch-531458aabdc4bfee)
    - [debug-batch brief](#debug-batch-brief-531458aabdc4bfee)
    - [debug-batch detailed](#debug-batch-detailed-531458aabdc4bfee)
- [Release](#release-531458aabdc4bfee)
    - [release brief](#release-brief-531458aabdc4bfee)
    - [release detailed](#release-detailed-531458aabdc4bfee)



<a name="debug-batch-531458aabdc4bfee">

# Debug-batch

<a name="debug-batch-brief-531458aabdc4bfee">

## debug-batch brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 13,661,498,361,007 | 10,621,579,156,095 | -3,039,919,204,912 | -22.25% :white_check_mark:
time.swift-driver.wall | 1347.9s | 1243.7s | -104.2s | -7.73% :white_check_mark:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
LLVM.NumLLVMBytesOutput | 494,009,138 | 493,842,012 | -167,126 | -0.03%
</details>

<a name="debug-batch-detailed-531458aabdc4bfee">

## debug-batch detailed
<details>
<summary>Regressed (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumSourceLinesPerSecond | 536,270 | 633,179 | 96,909 | 18.07% :no_entry:
Sema.UnderlyingTypeDeclsReferencedRequest | 1,539,720 | 1,557,479 | 17,759 | 1.15% :no_entry:
</details>
<details>
<summary>Improved (14)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumASTBytesAllocated | 13,535,471,797 | 11,875,122,523 | -1,660,349,274 | -12.27% :white_check_mark:
Driver.ChildrenMaxRSS | 37,308,805,120 | 34,784,794,624 | -2,524,010,496 | -6.77% :white_check_mark:
Driver.NumDriverPipePolls | 208,604 | 88,360 | -120,244 | -57.64% :white_check_mark:
Driver.NumDriverPipeReads | 216,370 | 94,050 | -122,320 | -56.53% :white_check_mark:
Frontend.MaxMallocUsage | 222,459,721,248 | 186,691,370,904 | -35,768,350,344 | -16.08% :white_check_mark:
Frontend.NumInstructionsExecuted | 13,661,498,361,007 | 10,621,579,156,095 | -3,039,919,204,912 | -22.25% :white_check_mark:
SILModule.NumSILOptFunctions | 678,858 | 671,411 | -7,447 | -1.1% :white_check_mark:
Sema.NumConformancesDeserialized | 1,965,163 | 1,936,201 | -28,962 | -1.47% :white_check_mark:
Sema.NumLazyGenericEnvironments | 3,152,773 | 3,118,584 | -34,189 | -1.08% :white_check_mark:
Sema.NumLazyIterableDeclContexts | 2,911,608 | 2,873,716 | -37,892 | -1.3% :white_check_mark:
Sema.NumTypesDeserialized | 7,174,677 | 6,957,585 | -217,092 | -3.03% :white_check_mark:
Sema.NumUnloadedLazyIterableDeclContexts | 2,019,234 | 1,982,340 | -36,894 | -1.83% :white_check_mark:
Sema.OverriddenDeclsRequest | 1,529,074 | 708,928 | -820,146 | -53.64% :white_check_mark:
Sema.SuperclassTypeRequest | 15,357 | 14,769 | -588 | -3.83% :white_check_mark:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (76)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumDecls | 33,401 | 33,401 | 0 | 0.0%
AST.NumDependencies | 90,765 | 90,767 | 2 | 0.0%
AST.NumImportedExternalDefinitions | 773,460 | 772,120 | -1,340 | -0.17%
AST.NumInfixOperators | 13,113 | 13,113 | 0 | 0.0%
AST.NumLinkLibraries | 0 | 0 | 0 | 0.0%
AST.NumLoadedModules | 103,611 | 103,611 | 0 | 0.0%
AST.NumLocalTypeDecls | 111 | 111 | 0 | 0.0%
AST.NumObjCMethods | 12,397 | 12,397 | 0 | 0.0%
AST.NumPostfixOperators | 14 | 14 | 0 | 0.0%
AST.NumPrecedenceGroups | 6,932 | 6,932 | 0 | 0.0%
AST.NumPrefixOperators | 60 | 60 | 0 | 0.0%
AST.NumReferencedDynamicNames | 101 | 101 | 0 | 0.0%
AST.NumReferencedMemberNames | 1,812,978 | 1,813,724 | 746 | 0.04%
AST.NumReferencedTopLevelNames | 115,653 | 115,653 | 0 | 0.0%
AST.NumSourceBuffers | 153,217 | 153,216 | -1 | -0.0%
AST.NumSourceLines | 1,193,359 | 1,193,359 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 2,546,056 | 2,541,148 | -4,908 | -0.19%
AST.NumUsedConformances | 101,130 | 101,126 | -4 | -0.0%
Driver.DriverDepCascadingDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingTopLevel | 0 | 0 | 0 | 0.0%
Driver.DriverDepDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepTopLevel | 0 | 0 | 0 | 0.0%
Driver.NumDriverJobsRun | 7,171 | 7,171 | 0 | 0.0%
Driver.NumDriverJobsSkipped | 0 | 0 | 0 | 0.0%
Driver.NumProcessFailures | 0 | 0 | 0 | 0.0%
Frontend.NumProcessFailures | 0 | 0 | 0 | 0.0%
IRModule.NumIRAliases | 59,044 | 59,044 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 1,746,666 | 1,746,321 | -345 | -0.02%
IRModule.NumIRComdatSymbols | 0 | 0 | 0 | 0.0%
IRModule.NumIRFunctions | 927,138 | 926,792 | -346 | -0.04%
IRModule.NumIRGlobals | 1,119,810 | 1,119,204 | -606 | -0.05%
IRModule.NumIRIFuncs | 0 | 0 | 0 | 0.0%
IRModule.NumIRInsts | 20,117,921 | 20,114,893 | -3,028 | -0.02%
IRModule.NumIRNamedMetaData | 35,398 | 35,398 | 0 | 0.0%
IRModule.NumIRValueSymbols | 1,815,366 | 1,814,414 | -952 | -0.05%
LLVM.NumLLVMBytesOutput | 494,009,138 | 493,842,012 | -167,126 | -0.03%
Parse.NumFunctionsParsed | 62,780 | 62,780 | 0 | 0.0%
SILModule.NumSILGenDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILGenFunctions | 841,617 | 841,272 | -345 | -0.04%
SILModule.NumSILGenGlobalVariables | 15,099 | 15,099 | 0 | 0.0%
SILModule.NumSILGenVtables | 3,778 | 3,778 | 0 | 0.0%
SILModule.NumSILGenWitnessTables | 18,544 | 18,544 | 0 | 0.0%
SILModule.NumSILOptDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILOptGlobalVariables | 15,511 | 15,511 | 0 | 0.0%
SILModule.NumSILOptVtables | 7,244 | 7,244 | 0 | 0.0%
SILModule.NumSILOptWitnessTables | 36,426 | 36,437 | 11 | 0.03%
Sema.AccessLevelRequest | 868,925 | 868,929 | 4 | 0.0%
Sema.DefaultAndMaxAccessLevelRequest | 22,137 | 22,137 | 0 | 0.0%
Sema.EnumRawTypeRequest | 7,723 | 7,723 | 0 | 0.0%
Sema.ExtendedNominalRequest | 1,497,660 | 1,492,231 | -5,429 | -0.36%
Sema.InheritedDeclsReferencedRequest | 52,767,604 | 52,768,332 | 728 | 0.0%
Sema.InheritedTypeRequest | 371,006 | 370,423 | -583 | -0.16%
Sema.IsDynamicRequest | 789,806 | 789,820 | 14 | 0.0%
Sema.IsObjCRequest | 651,277 | 651,291 | 14 | 0.0%
Sema.NamedLazyMemberLoadFailureCount | 14,102 | 14,102 | 0 | 0.0%
Sema.NamedLazyMemberLoadSuccessCount | 2,288,050 | 2,294,272 | 6,222 | 0.27%
Sema.NominalTypeLookupDirectCount | 15,121,686 | 15,145,576 | 23,890 | 0.16%
Sema.NumConstraintScopes | 7,332,893 | 7,331,854 | -1,039 | -0.01%
Sema.NumConstraintsConsideredForEdgeContraction | 13,181,633 | 13,180,279 | -1,354 | -0.01%
Sema.NumDeclsDeserialized | 16,289,852 | 16,201,682 | -88,170 | -0.54%
Sema.NumDeclsValidated | 891,624 | 891,629 | 5 | 0.0%
Sema.NumFunctionsTypechecked | 645,846 | 644,524 | -1,322 | -0.2%
Sema.NumGenericSignatureBuilders | 776,626 | 775,120 | -1,506 | -0.19%
Sema.NumLazyGenericEnvironmentsLoaded | 298,030 | 297,852 | -178 | -0.06%
Sema.NumTypesValidated | 608,995 | 609,000 | 5 | 0.0%
Sema.RequirementRequest | 22,769 | 22,769 | 0 | 0.0%
Sema.SelfBoundsFromWhereClauseRequest | 52,018 | 52,018 | 0 | 0.0%
Sema.SetterAccessLevelRequest | 54,206 | 54,206 | 0 | 0.0%
Sema.SuperclassDeclRequest | 41,713,211 | 41,793,459 | 80,248 | 0.19%
Sema.USRGenerationRequest | 175,720 | 175,720 | 0 | 0.0%
</details>

<a name="release-531458aabdc4bfee">

# Release

<a name="release-brief-531458aabdc4bfee">

## release brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
time.swift-driver.wall | 2496.8s | 2402.2s | -94.6s | -3.79% :white_check_mark:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 38,174,823,964,542 | 38,333,337,545,294 | 158,513,580,752 | 0.42%
LLVM.NumLLVMBytesOutput | 475,197,068 | 475,030,286 | -166,782 | -0.04%
</details>

<a name="release-detailed-531458aabdc4bfee">

## release detailed
<details>
<summary>Regressed (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Sema.NumConformancesDeserialized | 936,944 | 950,394 | 13,450 | 1.44% :no_entry:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (22)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumImportedExternalDefinitions | 140,898 | 140,898 | 0 | 0.0%
AST.NumLoadedModules | 6,613 | 6,613 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 462,818 | 462,731 | -87 | -0.02%
AST.NumUsedConformances | 104,005 | 104,001 | -4 | -0.0%
IRModule.NumIRBasicBlocks | 1,715,122 | 1,714,734 | -388 | -0.02%
IRModule.NumIRFunctions | 761,808 | 761,434 | -374 | -0.05%
IRModule.NumIRGlobals | 880,969 | 880,363 | -606 | -0.07%
IRModule.NumIRInsts | 14,734,924 | 14,734,137 | -787 | -0.01%
IRModule.NumIRValueSymbols | 1,516,110 | 1,515,130 | -980 | -0.06%
LLVM.NumLLVMBytesOutput | 475,197,068 | 475,030,286 | -166,782 | -0.04%
SILModule.NumSILGenFunctions | 323,195 | 322,846 | -349 | -0.11%
SILModule.NumSILOptFunctions | 448,293 | 447,171 | -1,122 | -0.25%
Sema.NumConstraintScopes | 7,253,920 | 7,252,797 | -1,123 | -0.02%
Sema.NumDeclsDeserialized | 2,920,597 | 2,915,862 | -4,735 | -0.16%
Sema.NumDeclsValidated | 417,299 | 417,299 | 0 | 0.0%
Sema.NumFunctionsTypechecked | 262,074 | 262,074 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 114,658 | 114,675 | 17 | 0.01%
Sema.NumLazyGenericEnvironments | 511,190 | 510,901 | -289 | -0.06%
Sema.NumLazyGenericEnvironmentsLoaded | 58,147 | 58,012 | -135 | -0.23%
Sema.NumLazyIterableDeclContexts | 342,446 | 342,243 | -203 | -0.06%
Sema.NumTypesDeserialized | 1,649,750 | 1,651,361 | 1,611 | 0.1%
Sema.NumTypesValidated | 187,234 | 187,234 | 0 | 0.0%
</details>

**Pull request merged**